pkg/probe/grpc/grpc.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	grpchealth "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"k8s.io/component-base/version"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/probe"
+)
+
+// Prober is an interface that defines the Probe function for doing GRPC readiness/liveness checks.
+type Prober interface {
+	Probe(host, service string, port int, timeout time.Duration, opts ...grpc.DialOption) (probe.Result, string, error)
+}
+
+type grpcProber struct {
+}
+
+// New Prober for execute grpc probe
+func New() Prober {
+	return grpcProber{}
+}
+
+// Probe executes a grpc call to check the liveness/readiness of container.
+// Returns the Result status, command output, and errors if any.
+// Only return non-nil error when service is unavailable and/or not implementing the interface
+// Otherwise result status is failed,BUT err is nil
+func (p grpcProber) Probe(host, service string, port int, timeout time.Duration, opts ...grpc.DialOption) (probe.Result, string, error) {
```

Comment:
ok, more comments to this method:

1. I'm mildly concern with the amount of error logs this method will produce as it logs error on EVERY error. It will be A LOT, especially for cases when grpc is simply not implemented.
2. Does the method needs to log errors at all? http probe only doing `V(4)` debug logging. If the method returns `err` or includes `err` in the result status, why do we need to log at all? I think the failure will be logged again when needed by the caller.
2. When you have if/else to differentiate error logs - the same differentiation will be appreciated in the error message returned. We need to make sure we are giving the best diagnostics for the failed probes to the customer without the need to dig into logs
3. `DeadlineExceeded` should not be returning `err`. Timeout is a "regular"/"expected" failure in the most cases. 
4. Failure to connect to the service in dial (if this is not a timeout) is likely a misconfiguration and should return `err`. What would be the reasons for it happening other than port is not being listened for?

This method is the most important piece of logic in this PR. Let's make it great! Thank you for working on it!

Reply:
1& 2 Yes, log is ALOT, and V(4) will be ok
3. Sure
4. Fix it
5. I think you are right, and fix it soon.